Control name filters at runtime

clementguidi opened the pull request:
These commits give the user full control over tracing filters at runtime. It
applies to the following filters:
 - opt-in filters set by `-F/--filter`
 - opt-out filters set by `-N/--notrace`
 - caller filters set by `-C/--caller-filter`

It allows the user to control the target process from the client as follows:
 - add or update an opt-in name filter
   - `$ uftrace -p PID -F func`
 - add or update an opt-out name filter
   - `$ uftrace -p PID -N func`
 - remove an opt-in or opt-out name filter
   - `$ uftrace -p PID -F func@clear`
 - add a caller filter
   - `$ uftrace -p PID -C func`
 - delete a caller filter
   - `$ uftrace -p PID -C func@clear`
 - set the pattern matching type
   - `$ uftrace -p PID --match <simple|regex|glob>`


 We also introduce an internal agent command `AGENT_CAPABILITIES`. It prompts
 the agent to indicate what features it supports.

namhyung commented on cmds/live.c:
```diff
@@ -340,9 +340,19 @@ static int forward_options(struct uftrace_opts *opts)
-	status = forward_option(sfd, capabilities, UFTRACE_AGENT_OPT_PATTERN, &opts->patt_type, sizeof(opts->patt_type));
-	if (status < 0)
-		goto close;
+	if (opts->filter) { /* provide a pattern type for options that need it */
```
`s/filter/patt_type/`?

clementguidi replied:
No, I mean to send the pattern type whenever the user uses a filter (or caller filter, trigger, args in later commits). In other words don't send the pattern type if the agent doesn't need it.

namhyung replied:
I see, thanks for the explanation.

namhyung commented on libmcount/mcount.c:
```diff
@@ -1785,6 +1785,20 @@ static void mcount_script_init(enum uftrace_pattern_type patt_type)
+static void agent_setup_filter(char *filter_str)
+{
+	if (agent_guard_filter()) {
+		usleep(100); /* grace period */
```
I'm not sure if it's enough, especially on large machines.

namhyung commented on libmcount/mcount.c:
```diff
@@ -870,7 +890,10 @@ enum filter_result mcount_entry_filter_check(struct mcount_thread_data *mtdp, un
 		return FILTER_OUT;
 
-	uftrace_match_filter(child, &mcount_triggers, tr);
+	if (!agent_run || agent_guard_filter()) {
```
I'm not sure if it's really safe.  Even with the full memory barrier, isn't it possible for concurrent readers and the agent can get the writer marker?

Also, I don't like the fact that it might miss the whole filters when agent makes the change.  Can we do it better?

clementguidi replied:
I agree this is not safe enough. There are a few ways we can tackle this.

### Disable tracing when agent is working

Our view on this is that if you are interacting with the agent to tweak tracing parameters, you do this for future events, because chances are the current calls are already over when the agent gets the instructions. So you may well skip instrumentation while the agent makes changes.

We can have the agent disable tracing by itself. This requires synchronizing.

We can have the user disabling tracing, and the agent would fail to apply filters if tracing is enabled. Without locks, this comes down to a grace period.

### Use read/write locks

We can introduce read/write locks, and skip instrumentation when data is read-protected. This requires a read lock in the fast path, which comes with an unwanted performance impact.

As discussed during our meeting in January, we could implement this solution in a separate libmcount build, e.g. `libmcount-sync.so`.

### Else

As for lockless algorithms, I don't think that would be simple to do.

namhyung replied:
I'm not happy with the both solutions.  Maybe we could have a simple RCU style algorithm to atomically update the whole filters and free the old copy a few seconds later.  Thoughts?

clementguidi replied:
OK. I thought about RCU and we should give it a try. Having a RCU rbtree with per-node RCU is out of reach IMO (quiet complex, and the code I can find is old. See [1] and [2]). Instead we may copy the whole tree. I'll try that.

[1] https://lists.lttng.org/pipermail/lttng-dev/2017-December/027640.html
[2] https://github.com/philip-w-howard/RP-Red-Black-Tree

clementguidi replied:
This is out of topic, but I'm curious: why use rbtrees in the first place? They better perform insertion/deletion than lookup, which is the opposite of what libmcount does. Is this from Ftrace?

namhyung replied:
Yeah, I also meant replacing the whole tree.

The rbtree was needed because it needs to check the address range.  Since the location of `mcount` call varies among functions, it couldn't check a specific offset from the function start.  Maybe we can reconsider a hash table if it only supports dynamic tracing because it can control the location.

namhyung reviewed (commented):
Thanks for your work!

clementguidi commented:
I rebased on #1678.

I need your view on this: I replaced the flags for filtering mode and location filtering mode to counters (count the registered filters instead of just IN/OUT flags). Those are global variables, and match the rbtree structure. Since we use RCU, we need to have a set of counters (same if we keep flags) to match the RCU copy and one for the original. **I want to have a structure with both the rbtree and the counters, and handle it with RCU** instead of just the tree. What do you think? Something as follows:

```
struct mcount_triggers_info {
    struct rb_root triggers;
    int filter_count;
    int loc_count;
    int caller_count;
}
```

Note: Tests fail on optimization level 2 and 3 because two functions in the test program get merged into a single one, I'll fix it.

namhyung commented on utils/filter.c:
```diff
@@ -45,8 +45,10 @@ static void print_trigger(struct uftrace_trigger *tr)
+		else if (tr->fmode == FILTER_MODE_OUT)
 			pr_dbg("\ttrigger: filter OUT\n");
+		else
+			pr_dbg("\ttrigger: filter NONE\n");
```
I think it should `tr->fmode` set when it has `TRIGGER_FL_FILTER` flag.

clementguidi replied:
Sorry, what do you mean? That the `NONE` value isn't valid when `TRIGGER_FL_FILTER` is set?

namhyung replied:
Right, it should not have `NONE` together with `TRIGGER_FL_FILTER`.

namhyung commented on utils/filter.c:
```diff
@@ -249,6 +251,11 @@ static void add_arg_spec(struct list_head *arg_list, struct uftrace_arg_spec *ar
 void update_trigger(struct uftrace_filter *filter, struct uftrace_trigger *tr, bool exact_match)
 {
+	if (tr->flags & TRIGGER_FL_CLEAR) {
+		filter->trigger.flags &= ~tr->flags;
```
Hmm.. what's the intended use case?  I think we can simply do `filter->triggers.flags = 0`, right?

clementguidi replied:
I only want to remove a given filter type. E.g. if you have a filter and an trigger set for a symbol, you can remove either the filter or the trigger (triggers are not implemented yet in the agent). Setting the flags to `0` clears all filters/triggers/args.

The following example would preserve the depth trigger on `func`.
```
$ uftrace record -F func -T func@depth=1
$ uftrace -p <pid> -F func@clear
```

namhyung replied:
Hmm.. ok.  Then `-T func@clear` would clear them all.

namhyung reviewed (commented):
I'm fine with having a struct to have all filter related settings.  By the way, why not renaming the test case to start from 281?

clementguidi commented on cmds/live.c:
```diff
@@ -340,10 +340,19 @@ static int forward_options(struct uftrace_opts *opts)
+
+	if (opts->filter) {
+		status = forward_option(sfd, capabilities, UFTRACE_AGENT_OPT_FILTER, opts->filter,
+					strlen(opts->filter));
```
Yes that makes more sense.

clementguidi commented on utils/filter.h:
```diff
@@ -40,6 +40,15 @@ enum trigger_flag {
+
+ * @FILTER_MODE_NONE is neutral and is only used for initialization in the
+ * location filter.
+ */
```
No problem!

clementguidi commented:
> why not renaming the test case to start from 281?

I'm using tests 281 and 282 in #1644, and I expected said PR to go through first.

clementguidi commented:
For now I'm keeping the `mcount_triggers` name but for the new structure (rbtree + counters). And I suggest we discuss naming separately, because it requires many modifications, especially in the code doc. What do you think?

```
struct mcount_triggers_info {
    struct rb_root triggers; /* the former 'mcount_triggers' rbtree */
    int filter_count; /* use counters instead of boolean mode */
    int caller_count;
    int loc_count;

struct mcount_triggers_info *mcount_triggers; /* RCU triggers data */
```

Regarding the nomenclature, the name `mcount_triggers` is ambiguous, as it refers to a rbtree with `struct uftrace_filter` nodes. And `struct uftrace_filter` itself contains a `struct uftrace_trigger` field. Eventually, the naming is not consistent, as in other places (e.g. `fstack.c`) the rbtree is named `filters`.

namhyung commented:
> I'm using tests 281 and 282 in https://github.com/namhyung/uftrace/pull/1644, and I expected said PR to go through first.

Ah, ok.  Let's do it first then :)

namhyung commented:
About the naming, I agree it's inconsistent.  It started with filters only, but later the triggers were added.  Now triggers are superset of filters with additional features.  To minimize the work, I think it's good to keep the `mcount_triggers`.  How about changing the name of tree as `root`?

clementguidi commented:
> How about changing the name of tree as `root`?

Yes sounds good.

namhyung reviewed (commented):
Can you please rebase this series?

I think we need to reconsider the filter mode/count change since it replaces the whole filter tree at once and no need to keep the number of filters.

clementguidi commented:
> I think we need to reconsider the filter mode/count change since it replaces the whole filter tree at once and no need to keep the number of filters.

I keep track of the number of filters to change the default filtering behavior on function entry. As long as there's a `-F` filter registered, functions are traced by default. I copy the counters so I don't have to parse the tree again. I don't understand how we can get rid of the mode/count?

clementguidi commented on libmcount/mcount.c:
```diff
@@ -1931,6 +1932,14 @@ static int agent_apply_option(int opt, void *value, size_t size, struct rb_root
+		opts.patt_type = *((int *)value);
+		if (opts.patt_type != mcount_filter_setting.ptype) {
+			mcount_filter_setting.ptype = opts.patt_type;
+			pr_dbg4("use pattern type %#x\n", opts.patt_type);
```
Yes, I didn't pay attention while rebasing.

clementguidi commented:
The counters are incremented when adding a filter, and decremented when removing one from the tree. As we copy the counters together with the tree, we don't need to parse the tree to get the count. Is that OK?

namhyung commented on utils/filter.c:
```diff
@@ -808,6 +822,10 @@ static const struct trigger_action_parser actions[] = {
+	{
+		"clear",
+		parse_clear_action,
+	},
```
Can you please update the documentation?

clementguidi replied:
I added a few lines in `uftrace-live.md` and `uftrace-record.md` to explain that the `@clear` suffix exists. I plan to add a reference to the agent section when I create it, where I will show examples. Is it OK for now? Or do you want to have examples in the doc in this PR?

namhyung replied:
You can do it in a separate PR.  Thanks!

namhyung reviewed (commented):
Looks good now.  I think it's ok to apply delta after changing the filters.

Btw, can you please add "agent" section in the uftrace live man page to explain how to use it with an example?

clementguidi commented:
> Btw, can you please add "agent" section in the uftrace live man page to explain how to use it with an example?

Yes, I'm sending another PR.

namhyung commented on tests/t284_agent_filter.py:
```diff
@@ -0,0 +1,89 @@
+
+from runtest import TestBase
+
+class TestCase(TestBase):
```
Again, please add a brief explanation for this test.

namhyung commented on tests/t285_agent_caller_filter.py:
```diff
@@ -0,0 +1,82 @@
+
+from runtest import TestBase
+
+class TestCase(TestBase):
```
Ditto.

namhyung reviewed (commented):
(Probably) last nitpick.  Please add a comment in the tests to describe what to check.

clementguidi commented:
> Please add a comment in the tests to describe what to check.

I'm not sure how detailed it needs to be, as the client commands in the test script are quite intuitive.

namhyung reviewed (approved):
Simple description like you added should be fine.  LGTM, thanks!
